File: app/src/main/java/com/soundcloud/android/offline/OfflineProperties.java
package com.soundcloud.android.offline;

import static java.util.Collections.emptyMap;

import com.google.auto.value.AutoValue;
import com.soundcloud.android.model.Urn;

import java.util.Map;

@AutoValue
public abstract class OfflineProperties {
    public static OfflineProperties from(Map<Urn, OfflineState> offlineEntitiesStates, OfflineState likedTracksState) {
        return new AutoValue_OfflineProperties(offlineEntitiesStates, likedTracksState);
    }

    public static OfflineProperties empty() {
        return from(emptyMap(), OfflineState.NOT_OFFLINE);
    }

    abstract Map<Urn, OfflineState> offlineEntitiesStates();

    public abstract OfflineState likedTracksState();

    public OfflineState state(Urn urn) {
        final Map<Urn, OfflineState> statesMap = offlineEntitiesStates();
        if (statesMap.containsKey(urn)) {
            return statesMap.get(urn);
        } else {
            return OfflineState.NOT_OFFLINE;
        }
    }
}

File: app/src/main/java/com/soundcloud/android/offline/OfflineProperties.kt
package com.soundcloud.android.offline

import com.soundcloud.android.model.Urn

data class OfflineProperties(val offlineEntitiesStates: Map<Urn, OfflineState> = mapOf<Urn, OfflineState>(),
                             val likedTracksState: OfflineState = OfflineState.NOT_OFFLINE) {

    fun state(urn: Urn) = offlineEntitiesStates.getOrElse(urn) { OfflineState.NOT_OFFLINE }

    fun apply(newState: OfflineProperties): OfflineProperties {
        val newOfflineEntitiesState = mutableMapOf<Urn, OfflineState>()
        newOfflineEntitiesState.putAll(offlineEntitiesStates)
        newOfflineEntitiesState.putAll(newState.offlineEntitiesStates)
        return OfflineProperties(newOfflineEntitiesState, newState.likedTracksState)
    }
}
